hostapd: fix reload frequency change patch

abhilashtuse opened the pull request:
Without this change ap+sta does not work. It updates current_mode
structure based on configured hw_mode received from wpa_supplicant.
Also prepare rates table after frequency selection.

Signed-off-by: Abhilash Tuse <Abhilash.Tuse@imgtec.com>

Ham22 commented:
This is a fix the previous patch 340 right, perhaps just best to fix it directly in the original patch? I'm not too familiar with this issue but seems like this is generic enough that you could upstream this and get their comments on it?

Or is there something us specific here?

abhilashtuse commented:
Yes. It fixes patch 340.

Commit:
hostapd: fix reload frequency change patch

When sta is configured, hostapd receives 'stop' and 'update' command from
wpa_supplicant. In the update command, hostapd gets sta parameters with
which it configures ap.

Problem is, with the default wireless configuration:
mode:11g freq:2.4GHz channel:1
If sta is connected to 5GHz network, then ap does not work. Ideally with
340-reload_freq_change.patch hostapd should reload the frequency changes
and start ap in 5GHz, but ap becomes invisible in the network.

This issue can be reproduced with following /etc/config/wireless:
config wifi-device  radio0
        option type     mac80211
        option channel  1
        option hwmode   11g
        option path     'virtual/uccp420/uccwlan'
        option htmode   'none'

config wifi-iface 'ap'
        option device 'radio0'
        option encryption 'none'
        option mode 'ap'
        option network 'ap'
        option ssid 'MyTestNet'
        option encryption none

config wifi-iface 'sta'
       option device radio0
       option network sta
       option mode sta
       option ssid TestNet-5G
       option encryption psk2
       option key 12345

This change updates current_mode structure based on configured hw_mode
received from wpa_supplicant. Also prepare rates table after frequency
selection.

Signed-off-by: Abhilash Tuse <Abhilash.Tuse@imgtec.com>
